In [29]:
import pandas as pd

# Load data using relative paths
brands = pd.read_csv('../data/raw/brands.csv')
orderlines = pd.read_csv('../data/raw/orderlines.csv')
orders = pd.read_csv('../data/raw/orders.csv')
products = pd.read_csv('../data/raw/products.csv')

# Create working copies
brands_df = brands.copy()
orderlines_df = orderlines.copy()
orders_df = orders.copy()
products_df = products.copy()

In [30]:
products_df_copy = products_df.copy()

In [31]:
products_df_copy.shape

(19326, 7)

In [32]:
products_df_copy.columns

Index(['sku', 'name', 'desc', 'price', 'promo_price', 'in_stock', 'type'], dtype='object')

In [33]:
nrows = products_df_copy.shape[0]
ncols = products_df_copy.shape[1]
print("The number of rows is", nrows)
print("The number of columns is", ncols)

The number of rows is 19326
The number of columns is 7


In [34]:
products_df_copy.size

135282

In [35]:
products_df_copy.shape[0] * products_df_copy.shape[1] == products_df_copy.size

True

In [36]:
products_df_copy.ndim

2

In [37]:
products_df_copy.head(20)

sku                                               name  \
0     RAI0007                  Silver Rain Design mStand Support   
1     APP0023                  Apple Mac Keyboard Keypad Spanish   
2     APP0025                   Mighty Mouse Apple Mouse for Mac   
3     APP0072      Apple Dock to USB Cable iPhone and iPod white   
4     KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5     APP0073     Apple Composite AV Cable iPhone and iPod white   
6     KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7     KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8   KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9     APP0100             Apple Adapter Mini Display Port to VGA   
10    APP0101              Apple Mini DisplayPort to DVI Adapter   
11    SEN0021  Sennheiser CX 300-II Precision headphones iPho...   
12    APP0111                     Apple USB Ethernet Adapter Mac   
13    APP0156              45W MagSafe charger Apple MacBook Air   
14    MOS0013         Adapter Moshi FireWire 400 to FireWire 800   
15    MOS0021               Clearguard Moshi MacBook Pro and Air   
16    APP0206  60W MagSafe charger Apple MacBook and MacBook ...   
17    HGD0007       Henge Docks Docking Station MacBook Pro 13 "   
18    MUV0009    Muvit Back Clear Case iPhone 4 / 4S Transparent   
19    APP0234                        Apple Dock Connector to VGA   

                                                 desc  price promo_price  \
0        Aluminum support compatible with all MacBook  59.99     499.899   
1           USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                              mouse Apple USB cable.     59     569.898   
3               IPhone dock and USB Cable Apple iPod.     25     229.997   
4   2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
5   IPhone and iPod AV Cable Dock to Composite Video.     45     420.003   
6   1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  18.99     146.471   
7        2GB RAM iMac with Intel Core 2 Duo (Penryn).  36.99     274.694   
8   RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...     74     669.904   
9   Adapter Mini Display Port to VGA MacBook and M...     35     330.003   
10  Adapter mini DisplayPort to DVI MacBook and Ma...     35     330.003   
11     Headphones iPhone iPad iPad 2 iPad 3 and iPod.  49.99     449.878   
12                       USB to Ethernet adapter Mac.     35     330.003   
13     Charger MacBook Air MagSafe 45 W (MC747Z / A).     89     849.904   
14           FireWire 400 adapter cable FireWire 800.     20     179.891   
15  Keyboard Protector MacBook Pro 13-inch Retina ...  24.95     249.889   
16  Charger MacBook and 13 inch MacBook MagSafe 60...     89     849.904   
17  Dock and support with multiple connections for...  69.99     699.864   
18                  Case iPhone 4 / 4S polycarbonate.   9.99       7.986   
19                         Dock Connector to VGA IOS.     35     350.005   

    in_stock      type  
0          1      8696  
1          0  13855401  
2          0      1387  
3          0      1230  
4          1      1364  
5          0      1230  
6          0      1364  
7          0      1364  
8          0      1364  
9          0      1325  
10         0      1325  
11         0      5384  
12         0      1334  
13         0  13005399  
14         0      1325  
15         0  13835403  
16         1  13005399  
17         0  12995397  
18         0  11865403  
19         0  13955395

In [38]:
products_df_copy.tail(20)

sku                                            name  \
19306  REP0429              iPhone Speaker lower repair 7 plus   
19307  REP0418                iPhone charging connector repair   
19308  REP0406                         battery repair iPad Air   
19309  ICA0115    Incase City Brief case MacBook Pro 13 "Black   
19310  REP0430                       iPad Speaker lower repair   
19311  REP0427                 Wi-Fi antenna repair Apple iPad   
19312  REP0424                    Input repair Headphones iPad   
19313  REP0421                  iPad charging connector repair   
19314  REP0416                        iPad front camera repair   
19315  REP0413                         repair rear camera iPad   
19316  REP0407                             iPad battery repair   
19317  REP0403                          iPad LCD screen repair   
19318  REP0402                        iPad touch screen repair   
19319  KNO0032  Knomo MacBook Pro Beauchamp Backpack 14 "Black   
19320  REP0076        repair Full screen iPad (1st generation)   
19321  BEL0376         Belkin Travel Support Apple Watch Black   
19322  THU0060    Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061     Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062      Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063    Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc      price  \
19306  Repair service including parts and labor for i...    599.906   
19307  Repair service including parts and labor for i...      49.99   
19308  Battery Repair Service includes parts and labo...  6.999.003   
19309      Incase bag to carry Apple MacBook Pro 13-inch      79.95   
19310  Repair service including parts and labor for iPad  6.999.003   
19311  Repair service including parts and labor for iPad  6.999.003   
19312  Repair service including parts and labor for iPad  6.999.003   
19313  Repair service including parts and labor for iPad  6.999.003   
19314  Repair service including parts and labor for iPad  6.999.003   
19315  Repair service including parts and labor for iPad  6.999.003   
19316  Repair service including parts and labor for iPad  6.999.003   
19317  Repair service including parts and labor for iPad     159.99   
19318  Repair service including parts and labor for iPad     139.99   
19319  Backpack thin nylon mesh internal compartment ...        179   
19320  Repair service including parts and labor for iPad     149.99   
19321  compact and portable stand vertically or horiz...      29.99   
19322  Backpack with capacity of 14 liter compartment...      69.95   
19323  Backpack with capacity of 14 liter compartment...      69.95   
19324  Backpack with capacity of 14 liter compartment...      69.95   
19325  Backpack with capacity of 14 liter compartment...      69.95   

      promo_price  in_stock      type  
19306     599.906         0  1,44E+11  
19307     499.899         0  1,44E+11  
19308       69.99         0  1,44E+11  
19309     739.903         1     10230  
19310       69.99         0  1,44E+11  
19311       69.99         0  1,44E+11  
19312       69.99         0  1,44E+11  
19313       69.99         0  1,44E+11  
19314       69.99         0  1,44E+11  
19315       69.99         0  1,44E+11  
19316       69.99         0  1,44E+11  
19317   1.599.898         0  1,44E+11  
19318   1.399.897         0  1,44E+11  
19319   1.699.905         1      1392  
19320   1.499.795         0  1,44E+11  
19321     269.903         1     12282  
19322     649.903         1      1392  
19323     649.903         1      1392  
19324     649.903         0      1392  
19325     649.903         1      1392

In [39]:
products_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


In [40]:
products_df_copy['price'].describe()

count     19280
unique     2690
top        3649
freq        466
Name: price, dtype: object

In [41]:
mask = products_df_copy['type'].duplicated(keep=False)
products_df_copy.loc[mask,:]

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19320  REP0076       repair Full screen iPad (1st generation)   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc   price promo_price  \
0           Aluminum support compatible with all MacBook   59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.      59     589.996   
2                                 mouse Apple USB cable.      59     569.898   
3                  IPhone dock and USB Cable Apple iPod.      25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...   34.99       31.99   
...                                                  ...     ...         ...   
19320  Repair service including parts and labor for iPad  149.99   1.499.795   
19322  Backpack with capacity of 14 liter compartment...   69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...   69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...   69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...   69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19320         0  1,44E+11  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[19318 rows x 7 columns]

In [42]:
products_df_copy.isna().sum()

sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64

In [43]:
desc_mask = products_df_copy['desc'].isna()
products_df_copy.loc[desc_mask, 'desc'] = products_df_copy.loc[desc_mask, 'name']
products_df_copy['desc'].isna().sum()

np.int64(0)

In [44]:
mask = products_df_copy['type'].isna()
products_df_copy.loc[mask,:]

sku                                               name  \
307      SAN0017       SanDisk Cruzer Edge USB 2.0 Flash Drive 16GB   
530      SAN0026    SanDisk Extreme Cruzer 16GB USB 3.0 Flash Drive   
792      APP0672                    Apple iPhone 5S 16GB Space Gray   
797      APP0673                        Apple iPhone 5S 16GB Silver   
798      APP0675                    Apple iPhone 5S 32GB Space Gray   
1193     APP0823                         Apple iPhone 6 16GB Silver   
1199     APP0829                    Apple iPhone 6 Plus 16GB Silver   
1200     APP0822                     Apple iPhone 6 16GB Space Gray   
1201     APP0825                     Apple iPhone 6 64GB Space Gray   
1202     APP0826                         Apple iPhone 6 64GB Silver   
1203     APP0828                Apple iPhone 6 Plus 16GB Space Gray   
1280     APP0856                          Apple iPhone 6 128GB Gold   
1281     APP0857                Apple iPhone 6 Plus 64GB Space Gray   
1282     APP0858                    Apple iPhone 6 Plus 64GB Silver   
1283     APP0860               Apple iPhone 6 Plus 128GB Space Gray   
1759     PAC0602               Pack Synology DS1815 + + 40TB WD Red   
1899     SAN0035       SanDisk Ultra 16GB SDHC Memory Card Class 10   
3016     APP1166                        Apple iPhone 6S 16GB Silver   
3055     ELG0014                           Elgato Video Capture Mac   
3090     APP1185                Apple iPhone 6S Plus 64GB Rose Gold   
3106     APP1167                    Apple iPhone 6S 16GB Space Gray   
3107     APP1169                     16GB Apple iPhone 6S Rose Gold   
3108     APP1170                        Apple iPhone 6S 64GB Silver   
3110     APP1171                    Apple iPhone 6S 64GB Space Gray   
3112     APP1172                          Apple iPhone 6S 64GB Gold   
3114     APP1173                     64GB Apple iPhone 6S Rose Gold   
3116     APP1178                   Apple iPhone 6S Plus 16GB Silver   
3117     APP1179               Apple iPhone 6S Plus 16GB Space Gray   
3118     APP1180                     Apple iPhone 6S Plus 16GB Gold   
3119     APP1181                Apple iPhone 6S Plus 16GB Rose Gold   
3120     APP1182                   Apple iPhone 6S Plus 64GB Silver   
3121     APP1183               Apple iPhone 6S Plus 64GB Space Gray   
3122     APP1184                     Apple iPhone 6S Plus 64GB Gold   
3123     APP1186                  Apple iPhone 6s Plus 128GB Silver   
3124     APP1187              Apple iPhone 6s Plus 128GB Space Gray   
3125     APP1188                    Apple iPhone 6s Plus 128GB Gold   
3126     APP1189               Apple iPhone 6s Plus 128GB Rose Gold   
3138     APP1168                          Apple iPhone 6S 16GB Gold   
10864    APP1202                    Apple iPad Wi-Fi 32GB Pro Plata   
11766    APP1478                           Apple iPhone 16GB Silver   
11767    APP1479                             Apple iPhone 16GB Gold   
11768    APP1480                        Apple iPhone 16GB Rose Gold   
11770    APP1482                           Apple iPhone 64GB Silver   
11771    APP1483                             Apple iPhone 64GB Gold   
15932    PAC2056  Second hand - Apple Mac mini Core 2 Duo 226Ghz...   
18741  GTE0093-A  Open - G-Technology G-DRIVE HDD 7200rpm 4TB Th...   
18822    DOD0010   Dodocool Sports Wireless Stereo Headphones Black   
18838    CAL0010  Station 3 Caldigit Thunderbolt Thunderbolt out...   
19092  SYN0173-A      Open - Synology DS218 + NAS server Mac and PC   
19185    RIN0017                                    Chime Bell Ring   

                                                    desc      price  \
307                         Pendrive USB 2.0 Mac and PC.          8   
530             USB 3.0 flash drive 16GB USB Mac and PC.         22   
792               New iPhone 5S 16G Libre (ME432Y / AB).  4.694.994   
797                New Free iPhone 5S 16GB (ME433Y / A).  4.090.042   
798                New Free iPhone 5

In [45]:
products_df_copy.info()
price_mask = products_df_copy["price"].str.count(r"\.") > 1
promo_price_mask = products_df_copy["promo_price"].str.count(r"\.") > 1
print(len(products_df_copy))
print(len(products_df_copy.loc[price_mask, :]))
print(len(products_df_copy.loc[promo_price_mask, :]))
print(
    f"Percentage of price_mask: {round(len(products_df_copy.loc[price_mask,:]) / len(products_df_copy)*100,3)}\nPercentage of promo_price_mask: {round(len(products_df_copy.loc[promo_price_mask,:]) / len(products_df_copy)*100,3)}\n"
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19326 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB
19326
431
12902
Percentage of price_mask: 2.23
Percentage of promo_price_mask: 66.76



In [46]:
products_df_copy.loc[price_mask,:].sample(20)

sku                                               name  \
16831  APP2302  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
19296  REP0422                    Headphone Input repair iPhone 7   
11558  PAC1269   Pack QNAP TS-251 | 4GB RAM | 8TB Seagate Desktop   
15917  PAC2021  Synology DS1517 + | 16GB RAM | 15TB (5x3TB) WD...   
15368  REP0319                    Connector Repair iPhone 6s load   
15349  REP0364                      iPad mini front camera repair   
16841  APP2315  Apple MacBook Pro 13 "Core i7 Touch Bar 35GHz ...   
15993  PAC1984  Synology DS1817 + | 16GB RAM | 80TB (8x10TB) S...   
13182  PAC1510                           QNAP TS-251A 8GB RAM NAS   
13228  PAC1524  Pack QNAP TS-451A NAS 8GB RAM l l 16TB (4x4TB)...   
15923  PAC2315  Synology DS1517 + NAS Server | 16GB RAM | 60TB...   
12869  PAC2026  Synology DS216 + II | Seagate 4TB IronWolf for...   
19174  RIN0008            Pro Video Camera Intercom Doorbell Ring   
12887  PAC2044  Synology DS216 + II | WD 4TB Network for Mac a...   
14565  PAC1558  SSD Crucial MX300 expansion kit 1TB + 16GB 160...   
12884  PAC2041  Synology DS216 + II | Seagate 4TB IronWolf for...   
1453   PAC0557  SSD expansion kit Samsung 850 Pro 256GB + 16GB...   
17490  APP2492                        Apple iPhone X 256GB Silver   
15331  REP0285  volume button repair on vibration control and ...   
11578  PAC1286  QNAP TS-451 Pack | 4GB RAM | Seagate Desktop 16TB   

                                                    desc       price  \
16831  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...  26.155.941   
19296  Repair service including parts and labor for i...   6.999.003   
11558  Pack NAS QNAP TS-251 4GB RAM + 8TB (2x4TB) Sea...   6.559.793   
15917  NAS server with 16GB of RAM and 15TB for Mac a...  16.298.392   
15368  Repair service including parts and labor for i...   6.999.003   
15349  Repair service including parts and labor for i...   6.999.003   
16841  New MacBook Pro 13 inch Touch Bar 35 GHz Core ...   2.855.594   
15993  NAS server with 16GB of RAM and 80TB for Mac a...  43.577.892   
13182  NAS with 8GB of RAM and USB port for direct ac...   4.156.689   
13228  NAS with 8GB RAM, and 4TB 4 hard drives for Ma...  13.568.689   
15923  NAS server with 16GB of RAM and 60TB for Mac a...  34.847.892   
12869  Synology DS216 + II with 2GB of RAM and 4TB (2...   5.939.496   
19174  Surveillance camera with microphone and an ang...   2.790.006   
12887  Synology DS216 + II with 8GB of RAM and 4TB (2...   7.059.696   
14565  SSD upgrade kit 1TB + 16GB 1600MHz RAM for Mac...   5.653.092   
12884  Synology DS216 + II with 8GB of RAM and 12TB (...  10.679.496   
1453   Expansion kit 256GB + SSD 16GB 1600Mhz RAM + D...   4.119.592   
17490               New Apple iPhone X 256GB Silver Free  13.290.011   
15331  Repair service including parts and labor for i...   6.999.003   
11578  QNAP TS-451 Pack + 4GB memory RAM + 16TB (4x4T...  10.709.569   

      promo_price  in_stock       type  
16831  26.155.941         0   1,02E+12  
19296       69.99         0   1,44E+11  
11558   5.429.899         0   12175397  
15917  14.977.465         0   12175397  
15368       69.99         0   1,44E+11  
15349       69.99         0   1,44E+11  
16841   2.855.594         0   1,02E+12  
15993  33.377.463         0   12175397  
13182   3.979.896         0   12175397  
13228  10.623.679         0   12175397  
15923  28.269.121         0   12175397  
12869   5.311.791         0   12175397  
19174   2.790.006         0       9094  
12887   6.731.787         0   12175397  
14565   4.675.839         1       1433  
12884     901.179         0   12175397  
1453    3.485.841         0       1433  
17490  13.290.011         1  113271716  
15331       69.99         0   1,44E+11  
11578   9.829.895         0   12175397

In [47]:
products_df_copy.loc[promo_price_mask,:].sample(20)

sku                                               name  \
2171   APP1949  Apple MacBook Pro Retina 13 "Core i5 29Ghz | 8...   
9982   PAC0980  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
8560   PAC1587  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
3894   APP1383  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 8GB ...   
10181  PAC1593  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
9657   PAC0964  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
4767   PAC1040  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   
1731   PAC1811           Pack Synology DS1515 + | WD 30TB Network   
7224   PAC1615  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB R...   
16596  APP2342  Apple MacBook Pro 15 "Core i7 Touch Bar 28GHz ...   
17895  PAC2271  Synology DS918 + NAS Server | 12GB | 8TB (4x2T...   
5006   PAC1044  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   
15915  PAC2019  Synology DS1517 + | 16GB RAM | 50TB (5x10TB) S...   
1573   PAC0583               Synology DS415 + Pack + 12TB Seagate   
7000   PAC1611  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   
3389   APP1375  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 8GB ...   
8436   PAC0924  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
3902   APP1385  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 8GB ...   
3496   APP1376  Apple iMac 27 "Core i7 Retina 5K 4GHz | 8GB | ...   
5713   PAC1056  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   

                                                    desc       price  \
2171   MacBook Pro 13-inch Retina display 29 Ghz i5 |...        2049   
9982   IMac desktop computer 27 inch 5K Retina i5 3.2...        3409   
8560   Desktop computer iMac 27-inch 3.2GHz Core i5 5...        3809   
3894   IMac desktop computer 27 inch 8GB RAM 2TB 5K R...        2929   
10181  Desktop computer iMac 27-inch 3.2GHz Core i5 5...        3409   
9657   IMac desktop computer 27 inch 5K Retina i5 3.2...        2809   
4767   IMac desktop computer 27 inch 5K Retina i5 3.3...        2869   
1731   DS1515 + NAS with 10TB hard drive Seagate Iron...     1336.94   
7224   IMac desktop computer 27 inch Retina 5K 4GHz C...        4369   
16596  New MacBook Pro 15-inch Touch Bar to 28GHz Cor...        3039   
17895  NAS server of the Plus Series for companies se...     1077.35   
5006   IMac desktop computer 27 inch 5K Retina i5 3.3...        3109   
15915       NAS with 16GB of RAM and 50TB for Mac and PC  29.847.892   
1573    Synology DS415 NAS Server Pack + + 12TB Seagate.     1022.99   
7000   Desktop computer iMac 27-inch 3.3GHz Core i5 5...        4309   
3389   IMac desktop computer 27 inch 5K Retina 8GB RA...        2749   
8436   IMac desktop computer 27 inch Retina 5K RAM 16...        2369   
3902   IMac desktop computer 27 inch 5K Retina 8GB RA...        3049   
3496   IMac desktop computer 27 inch 5K Retina 8GB RA...        3049   
5713   IMac desktop computer 27 inch 5K Retina i5 3.3...        3289   

      promo_price  in_stock      type  
2171   19.535.849         0      1282  
9982   29.799.904         0  5,74E+15  
8560   33.579.896         0      1282  
3894   27.895.848         0  5,74E+15  
10181  26.739.899         0  5,74E+15  
9657   25.369.901         0  5,74E+15  
4767   25.889.898         0  5,74E+15  
1731   12.229.627         0  12175397  
7224   34.699.896         0  5,74E+15  
16596  28.260.046         0  1,02E+12  
17895  10.203.676         0  12175397  
5006   28.169.901         0  5,74E+15  
15915  23.627.464         0  12175397  
1573    9.229.904         0  12175397  
7000   32.649.902         0  5,74E+15  
3389   24.755.947         0  5,74E+15  
8436   20.989.894         0      1282  
3902    2.903.585         0  5,74E+15  
3496    2.903.585         0  5,74E+15  
5713   29.889.904         0  5,74E+15

In [48]:
products_df_copy.loc[price_mask | promo_price_mask,:].sample(30)

sku                                               name  \
5495   PAC1052  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   
11368  SAN0138  SanDisk Ultra microSDXC 128GB Memory Card + SD...   
1547   PAC1827  Synology DS1515 + l l 8GB RAM 30TB (5x6TB) WD Red   
3577   APP1378  Apple iMac 27 "Core i7 Retina 5K 4GHz | 8GB | ...   
255    PAC1241  Apple MacBook Pro 133 "i7 29GHz | RAM 16GB | 5...   
6554   PAC1071  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   
8062   APP1269  Apple iMac 21.5 "Core i7 33 GHz Retina display...   
17385  OWC0259                     OWC Mercury Electra 6G SSD 1TB   
9091   APP1255  Apple iMac 21.5 "Core i5 2.8GHz | 8GB | 1TB HD...   
12309  DLL0041       Dell P2317H Monitor 23 "IPS DP 1.2 pivotably   
6475   PAC1070  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 32GB...   
16851  GLY0019  Glyph Blackbox Plus External Hard Drive SSD 51...   
1725   PAC1528             Synology DS1515 + NAS Server | 8GB RAM   
16338  APP2269  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
8246   PAC1579  Apple iMac 21.5 "Core i7 3.3GHz 4K Retina | 8G...   
4139   APP1389  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 8GB ...   
8783   APP1237  Apple iMac 215 "Core i5 16GHz | 16GB | 1TB Fus...   
16892  PAC1944  Apple Mac Pro 35GHz 6 cores | 64GB RAM | 256GB...   
14187  APP1891  Apple MacBook Pro 15 "Core i7 Touch Bar 29GHz ...   
7876   PAC1626  Apple iMac 27 "Core i7 Retina 5K 4GHz | 8GB RA...   
1383   APP1928  Apple Mac mini Core i7 3GHz | 16GB RAM | 1TB F...   
9205   PAC1569  Apple iMac 21.5 "Core i5 2.8GHz | 8GB RAM | 1T...   
5462   PAC1052  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 16GB...   
14795  PAC1726  Pack QNAP TS-451A NAS 8GB RAM l l 32TB (4x8TB)...   
7039   PAC1612  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 32GB...   
3859   APP1383  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 8GB ...   
6623   PAC1238  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   
11390  PAC1836              Synology DS416J | 32TB (4x8TB) WD Red   
7725   PAC1624  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 32GB...   
10160  PAC1592  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 8GB ...   

                                                    desc      price  \
5495   IMac desktop computer 27 inch 5K Retina i5 3.3...       3169   
11368  And adapter card memory 128 GB microSDXC Class...  5.028.034   
1547   NAS Synology DS1815 + 8GB memory and 30TB WD R...    2380.97   
3577   IMac desktop computer 27 inch 8GB RAM 256GB Re...       2929   
255    Apple MacBook Pro 133 inches (MD101Y / A) with...       2399   
6554   IMac desktop computer 27 inch 5K Retina 4GHz i...       4669   
8062   Apple iMac desktop Core i7 21.5-inch Retina di...       2329   
17385                  1TB HDD SATA III SSD 7mm for Mac.     435.99   
9091   PC 215 inch iMac i5 2.8GHz 8GB RAM 1TB HDD (MK...       1678   
12309                          Pivotable Monitor for Mac     231.11   
6475   IMac desktop computer 27 inch 5K Retina i5 3.3...       4369   
16851  512GB SSD external drive with USB-C connection...     269.99   
1725   NAS with 8GB RAM and 5 bays compatible with SA...     885.97   
16338  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...       2869   
8246   Desktop computer iMac 21.5-inch 3.3GHz Core i7...       3409   
4139   IMac desktop computer 27 inch 8GB RAM 512GB Re...       3169   
8783   PC 215 inch iMac RAM 16GB 1TB Fusion (MK142Y /...       1788   
16892  New Mac Pro with 64GB of RAM and two 6-core GP...       5129   
14187  New MacBook Pro 15-inch Core i7 Touch Bar 29Gh...       4999   
7876   IMac desktop computer 27 inch Retina 5K 4GHz C...       4669   
1383   Apple desktop Mac mini Core i7 3GHz | 8GB RAM ...       1579   
9205   Desktop computer iMac 21.5-inch 2.8GHz Core i5...       2249   
5462   IMac desktop computer 27 inch 5K Retina i5 3.3...       3169   
14795  NAS with 16GB of RAM and two hard drives for M...    1432.47   
7039   Desktop computer iMac 27-inch 3.3GHz Core i5 5...       4789   
3859   IMac desktop compu

In [49]:
products_df_copy.info()
mask_comma = products_df_copy["type"].str.contains(r"\,")
mask_plus = products_df_copy["type"].str.contains(r"\+")
type_mask = mask_comma & mask_plus
print(len(products_df_copy))
print(len(products_df_copy.loc[type_mask, :]))
print(
    f"Percentage of type: {round(len(products_df_copy.loc[type_mask,:]) / len(products_df_copy)*100,3)}"
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19326 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB
19326
7187
Percentage of type: 37.188


In [50]:
print(f'there are {len(products_df_copy.loc[(price_mask | promo_price_mask) & type_mask, :])} rows with scientific types and price problems')
print(f'there are {len(products_df_copy.loc[(price_mask | promo_price_mask) & ~(type_mask), :])} rows without scientific types and price problems')

there are 6672 rows with scientific types and price problems
there are 6331 rows without scientific types and price problems


In [51]:
orderlines_df.head()
merged_df = products_df_copy.merge(orderlines_df, on='sku')
merged_df.sample(20)

sku                                               name  \
2143    SEA0039  Seagate Barracuda 2TB 35 "SATA hard drive Mac ...   
323137  APP2429  Apple Macbook 12 "Core i7 14GHz | 8GB | 256GB ...   
121263  APP1198  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 8GB ...   
337195  TUC0339  Tucano Elements Second Skin MacBook Pro Case /...   
2245    SEA0039  Seagate Barracuda 2TB 35 "SATA hard drive Mac ...   
253209  TUC0298  Top Tucano Second Skin Case Macbook Pro 13 "(l...   
218273  APP1685  Adapter Apple Lightning to headphone jack 3.5 ...   
57786   NTE0082        NewerTech NuCube stand upright for Mac Mini   
373163  APP2369  Apple Macbook Pro 13 "Core i5 2.3GHz | 8GB | 1...   
118718  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
333008  BEL0303  Belkin Suit Line Backpack for MacBook Pro 13 "...   
143335  PAC1600  Apple iMac 27 "Core i7 Retina 5K 4Ghz | RAM 32...   
164456  MIC0036             Office 365 Personal 1 license 1 M ± or   
36238   PLA0014  Plantronics BackBeat GO 2 Bluetooth Headset Black   
198932  LAC0186               1TB LaCie Rugged USB 3.0 Thunderbolt   
102646  PAC1046  Apple iMac 27 "Core i5 3.3GHz Retina 5K | 32GB...   
168226  MIC0036             Office 365 Personal 1 license 1 M ± or   
61784   APP0958  Apple MacBook Pro Retina 13 "i5 27 Ghz | 8GB R...   
266039  APP1972                        Apple iPhone 32GB Rose Gold   
209054  ZAG0028  Zagg keyboard Messenger Case iPad Air / Air 2 ...   

                                                     desc   price promo_price  \
2143     Internal Hard Drive 2TB Mac and PC (ST2000DM006)   89.99     639.945   
323137  New MacBook Retina Display 12-inch Core i7 8GB...    1799   1.691.004   
121263  IMac desktop computer 27 inch 5K Retina 8GB RA...    2129  18.555.943   
337195  Neoprene 4mm thick velvet interior and MacBook...    29.9     119.899   
2245     Internal Hard Drive 2TB Mac and PC (ST2000DM006)   89.99     639.945   
253209  Protective cover with reinforced zipper and ad...    34.9     299.899   
218273  Lightning adapter for headphones or earphones ...       9           9   
57786    Protective stand for Mac Mini from 2010 to 2014.   60.99      44.989   
373163  New MacBook Pro 13-inch Core i7 23 GHz with 8G...    2499  23.490.045   
118718  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...     NaN     1305.59   
333008  Lightweight backpack with durable material and...   57.99     499.899   
143335  IMac desktop computer 27 inch Retina 5K 4GHz C...    3769  31.089.897   
164456  Personal license with subscription of 365 days...    69.9     689.906   
36238   Wireless headphones with lightweight design Bl...   89.99     499.899   
198932  Portable Hard Disk Thunderbolt and USB 3.0 Mac...  159.99   1.299.044   
102646  IMac desktop computer 27 inch 5K Retina i5 3.3...    3589  30.249.903   
168226  Personal license with subscription of 365 days...    69.9     689.906   
61784   New MacBook Pro 13-inch Retina screen i5 128GB...    1449     1305.59   
266039       New iPhone SE with 4-inch screen 32 free GB.     419   3.589.997   
209054  skinny sleeve with bluetooth keyboard and adju...   69.99     459.897   

        in_stock      type       id  id_order  product_id  product_quantity  \
2143           1  12655397  1210721    337878           0                 1   
323137         0  5,39E+11  1564678    492388           0                 1   
121263         0      1282  1195819    331053           0                 1   
337195         0  13835403  1393675    418385           0                 1   
2245           1  12655397  1364429    405897           0                 8   
253209         1  13835403  1233843    348998           0                 1   
218273         1      1230  1193751    330145           0                 1   
57786          0      8696  1358778    403157           0                 1   
373163         0  1,02E+12  1543617    484296           0                 1   
118718         0      1282  1262634    362880    

In [52]:
merged_df.loc[merged_df["promo_price"].str.count(r"\.") > 1,:].sample(20)

sku                                               name  \
47528     APP0899  Apple Mac mini Core i5 14GHz | 4GB RAM | 500GB...   
346164    APP2490                         Apple iPhone 64GB X Silver   
138059    PAC0962  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 32GB...   
132912    PAC0960  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
202476    CRU0051                       Crucial MX300 525GB SSD Disk   
153279    PAC1162             Synology Pack DS216SE | WD 4TB Network   
231464    APP1854  Apple MacBook Pro 13 "with Touch Bar 33GHz Cor...   
370450  TRA0029-A  Open - Transcend JetDrive Lite 130 256GB MacBo...   
263365    APP1966                           Apple iPhone 32GB Silver   
321141    APP2148         Apple iPad Pro 10.5 "Wi-Fi 64GB Space Gray   
345284    APP2489                     Apple iPhone 64GB Space Gray X   
111879    PAC1071  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   
142357    PAC1596  Apple iMac 27 "Core i5 3.2GHz Retina 5K | 16GB...   
26891     APP0662                     IMac AppleCare Protection Plan   
288418    PAC2148  Apple iMac 27 "Core i7 Retina 5K 42Ghz | 64GB ...   
334592    AP20264           Like new - Apple iPhone 7 Plus 128GB RED   
153394    PAC1170              Synology DS416 Pack | WD 12TB Network   
286462    APP2216  Apple iMac 27 "Core i5 3.4GHz Retina 5K | 8GB ...   
206454    SYN0148           Synology DS416slim NAS server Mac and PC   
105363    PAC1055  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   

                                                     desc        price  \
47528   PC Mac mini Core i5 4GB 500GB 14GHz (MGEM2YP /...       555.59   
346164                New Apple iPhone 64GB Free Silver X  115.900.092   
138059  IMac desktop computer 27 inch 5K Retina i5 3.2...         3049   
132912  IMac desktop computer 27 inch 5K Retina i5 3.2...         2569   
202476  SSD 525GB SATA Hard Disk 3.0 (6Gb / s) for Mac...       158.66   
153279  Synology Pack DS216SE + 4TB (2x2TB) Network WD...       337.99   
231464  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...         3279   
370450  MLC memory card reconditioned 256GB Macbook 13...          182   
263365       New iPhone SE with 4-inch screen 32 free GB.          419   
321141                            Pro New iPad Wi-Fi 64GB          729   
345284                             New Free iPhone 64GB X         1159   
111879  IMac desktop computer 27 inch 5K Retina 4GHz i...         4669   
142357  Desktop computer iMac 27-inch 3.2GHz Core i5 5...         4009   
26891              Apple Care extended warranty for iMac.          179   
288418  IMac desktop computer 27 inch 5K Retina 64GB R...         4519   
334592  Refurbished iPhone 7 Plus Color RED Edition wi...          889   
153394  Synology DS416 pack of 12TB (4x3TB) Network WD...       901.99   
286462  IMac desktop computer 27 inch 5K Retina 8GB RA...         2219   
206454  8TB NAS server storage core 1.0 GHz dual-frequ...       313.99   
105363  IMac desktop computer 27 inch 5K Retina 4GHz i...         3949   

       promo_price  in_stock       type       id  id_order  product_id  \
47528    5.220.049         0       1282  1138715    306336           0   
346164  11.590.009         1  113271716  1544194    484528           0   
138059  28.660.045         0   5,74E+15  1597440    504902           0   
132912  24.150.045         0   5,74E+15  1226863    345621           0   
202476   1.449.943         0   12215397  1361652    404546           0   
153279   2.971.784         1   12175397  1172811    320638           0   
231464  29.655.951         0       2158  1240019    351936           0   
370450   1.736.637         0   57445397  1574576    496303           0   
263365   3.589.997         1   51601716  1385561    416193           0   
321141   7.070.018         0  106431714  1314713    387854           0   
345284  11.240.005         1  113271716  1647948    526286           0   
111879  40.519.899         0   5,74E+15  1237132    350506

In [53]:
merged_df.loc[merged_df["unit_price"].str.count(r"\.") == 1,:].sample(20)

sku                                               name  \
325178    APP2072  Apple MacBook Air 13 "Core i5 18GHz | 8GB RAM ...   
350058    APP2501  Apple Watch GPS 42mm Case Series 3 Aluminum Sp...   
20551     DLK0032                   D-Link GO-SW-5G Gigabit Switch 5   
11440     APP0410          Apple Thunderbolt to FireWire 800 adapter   
219354    TUC0281  Tucano SOTTILE UltraSlim Case iPhone 8/7 Plus ...   
72880     APP0700  Apple 85W MagSafe 2 charger MacBook Pro screen...   
6252      WDT0140           Red 2TB WD 35 "Mac PC hard drive and NAS   
255592    WAC0222           South Intuos Pro Paper M Graphics Tablet   
333656  LIF0113-A  Open - LifeProof Nood Submersible iPhone Case ...   
319018    APP2117        Apple iPad Pro 12.9 "Wi-Fi 256GB Space Gray   
218676    APP1776  Smart Battery Apple Battery Case iPhone 6s Cha...   
271940    SAT0054     Satechi Hub Multiport USB-C 4K HDMI Space Gray   
90744     APP1190                   IPhone AppleCare Protection Plan   
72251     BEZ0212  Be.ez LA robe Case Graphite MacBook Pro 15 "(L...   
149393    APP1216                             Apple Magic Trackpad 2   
68771   SEA0047-A  Open - 4TB Seagate Barracuda 35 "SATA 5900rpm ...   
209840    SHE0044              SwitchEasy iPhone Case 0.35 8/7 White   
66670     REP0239                            battery repair iPhone 6   
165039    MIC0036             Office 365 Personal 1 license 1 M ± or   
204667    NTE0062  NewerTech NuPower Battery 65W MacBook Pro 13 "...   

                                                     desc    price  \
325178  laptop MacBook Air 13 inch i5 18GHz 8GB RAM an...  1105.59   
350058  42mm Apple Watch series 3 with GPS altimeter a...      399   
20551        Dlink Switch 5 port Gigabyte for Mac and PC.    19.99   
11440     Adapter Thunderbolt to Firewire 800 connection.       35   
219354  transparent and flexible cover with 05mm thick...      9.9   
72880   Apple MagSafe 2 Charger for MacBook Pro 15-inc...       89   
6252    Western Digital hard drive designed for NAS 2T...       99   
255592  Medium graphics tablet lets you design on pape...    429.9   
333656  waterproof and resistant to extreme conditions...    89.99   
319018                 New iPad Pro 12.9 inch Wi-Fi 256GB     1069   
218676                   Battery Case for iPhone 6s and 6      119   
271940  Aluminum adapter with USB Type-C HDMI 4K three...    89.99   
90744                Apple Care extended warranty iPhone.       70   
72251   flexible sheath with full protection for MacBo...    29.99   
149393                 Apple Wireless Bluetooth Trackpad.      149   
68771   internal hard drive with Mac and PC open to 4T...   132.99   
209840            Ultra Thin Case for iPhone 8/7 in white     9.99   
66670   Repair service including parts and labor for i...    59.99   
165039  Personal license with subscription of 365 days...     69.9   
204667  65W internal battery for MacBook Pro 13 inch 2...   131.99   

       promo_price  in_stock      type       id  id_order  product_id  \
325178   9.690.043         1  2,17E+11  1619622    514139           0   
350058   3.909.994         1  24895185  1530281    478646           0   
20551      169.896         0      1334  1192533    329612           0   
11440      330.003         1      1325  1317486    389210           0   
219354      79.896         0  11865403  1249779    356700           0   
72880      649.903         1  13005399  1316223    388605           0   
6252       805.945         1  12655397  1472660    453746           0   
255592   3.969.889         0      1405  1249056    356409           0   
333656     549.595         0      1298  1286626    374606           0   
319018  10.370.015         0  51861714  1270896    367059           0   
218676   1.099.902         0  5,49E+11  1306326    383900           0   
271940       69.99         1  12585395  1470793    453031           0   
90744      609.997         1      1231  1555350    488826           0   
72251   

In [54]:
merged_df['new_unit_price'] = merged_df['unit_price']
mask = merged_df["unit_price"].str.count(r"\.") > 1
merged_df.loc[mask, "new_unit_price"] = merged_df.loc[mask, "unit_price"].str.replace(pat=r'\.',repl='',n=1,regex=True)
merged_df["new_unit_price"] = pd.to_numeric(merged_df["new_unit_price"])
merged_df = merged_df.loc[:,['sku', 'name', 'desc', 'price', 'promo_price', 'new_unit_price']]


In [55]:
price_exploration_df = merged_df.copy()
price_mask = price_exploration_df["price"].str.count(r"\.") > 1
price_exploration_df['price_replace'] = price_exploration_df.loc[price_mask, "price"].str.replace(pat=r'\.',repl='',regex=True)
price_exploration_df

sku                                      name  \
0       RAI0007         Silver Rain Design mStand Support   
1       RAI0007         Silver Rain Design mStand Support   
2       RAI0007         Silver Rain Design mStand Support   
3       RAI0007         Silver Rain Design mStand Support   
4       RAI0007         Silver Rain Design mStand Support   
...         ...                                       ...   
378392  REP0076  repair Full screen iPad (1st generation)   
378393  REP0076  repair Full screen iPad (1st generation)   
378394  REP0076  repair Full screen iPad (1st generation)   
378395  REP0076  repair Full screen iPad (1st generation)   
378396  REP0076  repair Full screen iPad (1st generation)   

                                                     desc   price promo_price  \
0            Aluminum support compatible with all MacBook   59.99     499.899   
1            Aluminum support compatible with all MacBook   59.99     499.899   
2            Aluminum support compatible with all MacBook   59.99     499.899   
3            Aluminum support compatible with all MacBook   59.99     499.899   
4            Aluminum support compatible with all MacBook   59.99     499.899   
...                                                   ...     ...         ...   
378392  Repair service including parts and labor for iPad  149.99   1.499.795   
378393  Repair service including parts and labor for iPad  149.99   1.499.795   
378394  Repair service including parts and labor for iPad  149.99   1.499.795   
378395  Repair service including parts and labor for iPad  149.99   1.499.795   
378396  Repair service including parts and labor for iPad  149.99   1.499.795   

        new_unit_price price_replace  
0                54.99           NaN  
1                49.99           NaN  
2                49.99           NaN  
3                49.99           NaN  
4                49.99           NaN  
...                ...           ...  
378392          149.99           NaN  
378393          149.99           NaN  
378394          149.99           NaN  
378395          149.98           NaN  
378396          149.98           NaN  

[378397 rows x 7 columns]

In [56]:
# Promo Price Pattern Analysis
# Note: new_unit_price was already created in the previous cell

# Step 1: Verify new_unit_price exists
print("\n1. Checking new_unit_price column...")
print(f"   new_unit_price column exists: {'new_unit_price' in merged_df.columns}")
print(f"   Sample values: {merged_df['new_unit_price'].head(3).tolist()}")

# Step 2: Identify problematic rows
print("\n2. Identifying problematic rows...")
price_mask = merged_df['price'].str.count(r'\.') > 1
promo_price_mask = merged_df['promo_price'].str.count(r'\.') > 1
print(f"   - Price problems: {price_mask.sum()} ({price_mask.sum()/len(merged_df)*100:.1f}%)")
print(f"   - Promo price problems: {promo_price_mask.sum()} ({promo_price_mask.sum()/len(merged_df)*100:.1f}%)")

# Step 3: Analyze promo_price patterns
# Test 1: Remove first dot
problematic_promo = merged_df[promo_price_mask].copy()
problematic_promo['test_remove_first_dot'] = problematic_promo['promo_price'].str.replace(r'\.', '', n=1, regex=True)
problematic_promo['test_numeric'] = pd.to_numeric(problematic_promo['test_remove_first_dot'], errors='coerce')
problematic_promo['ratio_to_unit'] = problematic_promo['test_numeric'] / problematic_promo['new_unit_price']

# How many are "reasonable"? (Promo price should be 0.5-1.5x unit_price)
reasonable = (problematic_promo['ratio_to_unit'] >= 0.5) & (problematic_promo['ratio_to_unit'] <= 1.5)
print(f"\n   TEST 1: Remove first dot")
print(f"   -> {reasonable.sum()} / {len(problematic_promo)} rows have reasonable ratio (0.5-1.5x)")
print(f"   -> Success rate: {reasonable.sum()/len(problematic_promo)*100:.1f}%")

# Show examples
print("\n   Examples:")
print(problematic_promo[['promo_price', 'new_unit_price', 'test_remove_first_dot', 'ratio_to_unit']].head(10))

# Test 2: Remove ALL dots, then divide by 100
problematic_promo['test_remove_all_dots'] = problematic_promo['promo_price'].str.replace(r'\.', '', regex=True)
problematic_promo['test_numeric_2'] = pd.to_numeric(problematic_promo['test_remove_all_dots'], errors='coerce') / 100
problematic_promo['ratio_to_unit_2'] = problematic_promo['test_numeric_2'] / problematic_promo['new_unit_price']

reasonable_2 = (problematic_promo['ratio_to_unit_2'] >= 0.5) & (problematic_promo['ratio_to_unit_2'] <= 1.5)
print(f"\n   TEST 2: Remove all dots, divide by 100")
print(f"   -> {reasonable_2.sum()} / {len(problematic_promo)} rows have reasonable ratio")
print(f"   -> Success rate: {reasonable_2.sum()/len(problematic_promo)*100:.1f}%")

# Step 4: Specific Pattern Recognition
print("\n4. Specific pattern analysis...")
print("-"*80)

# Pattern 1: X.XXX.XXX (e.g., 7.640.001)
pattern_1 = problematic_promo[problematic_promo['promo_price'].str.match(r'^\d\.\d{3}\.\d{3}$', na=False)].copy()
print(f"\n   PATTERN: X.XXX.XXX (like 7.640.001)")
print(f"   -> Found {len(pattern_1)} rows")
if len(pattern_1) > 0:
    pattern_1['fixed'] = pd.to_numeric(pattern_1['promo_price'].str.replace(r'\.', '', n=1, regex=True))
    pattern_1['ratio'] = pattern_1['fixed'] / pattern_1['new_unit_price']
    reasonable = (pattern_1['ratio'] >= 0.5) & (pattern_1['ratio'] <= 1.5)
    print(f"   -> {reasonable.sum()} are reasonable ({reasonable.sum()/len(pattern_1)*100:.1f}%)")
    print("\n   Sample:")
    print(pattern_1[['promo_price', 'new_unit_price', 'fixed', 'ratio']].head(5))

# Pattern 2: XX.XXX.XXX (e.g., 11.590.009)
pattern_2 = problematic_promo[problematic_promo['promo_price'].str.match(r'^\d{2}\.\d{3}\.\d{3}$', na=False)].copy()
print(f"\n   PATTERN: XX.XXX.XXX (like 11.590.009)")
print(f"   -> Found {len(pattern_2)} rows")
if len(pattern_2) > 0:
    pattern_2['fixed'] = pd.to_numeric(pattern_2['promo_price'].str.replace(r'\.', '', n=1, regex=True))
    pattern_2['ratio'] = pattern_2['fixed'] / pattern_2['new_unit_price']
    reasonable = (pattern_2['ratio'] >= 0.5) & (pattern_2['ratio'] <= 1.5)
    print(f"   -> {reasonable.sum()} are reasonable ({reasonable.sum()/len(pattern_2)*100:.1f}%)")
    print("\n   Sample:")
    print(pattern_2[['promo_price', 'new_unit_price', 'fixed', 'ratio']].head(5))

# Pattern 3: XXX.XXX.XXX (e.g., 115.900.092)
pattern_3 = problematic_promo[problematic_promo['promo_price'].str.match(r'^\d{3}\.\d{3}\.\d{3}$', na=False)].copy()
print(f"\n   PATTERN: XXX.XXX.XXX (like 115.900.092)")
print(f"   -> Found {len(pattern_3)} rows")
if len(pattern_3) > 0:
    pattern_3['fixed'] = pd.to_numeric(pattern_3['promo_price'].str.replace(r'\.', '', n=1, regex=True))
    pattern_3['ratio'] = pattern_3['fixed'] / pattern_3['new_unit_price']
    reasonable = (pattern_3['ratio'] >= 0.5) & (pattern_3['ratio'] <= 1.5)
    print(f"   -> {reasonable.sum()} are reasonable ({reasonable.sum()/len(pattern_3)*100:.1f}%)")
    print("\n   Sample:")
    print(pattern_3[['promo_price', 'new_unit_price', 'fixed', 'ratio']].head(5))

# Step 5: Summary
print("\n" + "="*80)
print("SUMMARY & RECOMMENDATION")
print("="*80)
print("\nBased on this analysis, the best strategy appears to be:")
print("1. Remove the FIRST dot from promo_price values with multiple dots")
print("2. This gives reasonable ratios for the majority of cases")
print("\nHowever, this won't fix 100% of cases. You'll need to:")
print("- Create a 'cleaned_promo_price' column")
print("- Flag rows that still look suspicious after cleaning")
print("- Decide how to handle the unfixable rows in your analysis")

# Recommended Fix
print("\n" + "="*80)
print("RECOMMENDED FIX")
print("="*80)
print("\nmerged_df['cleaned_promo_price'] = merged_df['promo_price'].copy()")
print("mask = merged_df['promo_price'].str.count(r'\\.') > 1")
print("merged_df.loc[mask, 'cleaned_promo_price'] = merged_df.loc[mask, 'promo_price'].str.replace(r'\\.', '', n=1, regex=True)")
print("merged_df['cleaned_promo_price'] = pd.to_numeric(merged_df['cleaned_promo_price'], errors='coerce')")


1. Checking new_unit_price column...
   new_unit_price column exists: True
   Sample values: [54.99, 49.99, 49.99]

2. Identifying problematic rows...
   - Price problems: 4195 (1.1%)
   - Promo price problems: 200095 (52.9%)

   TEST 1: Remove first dot
   -> 9791 / 200095 rows have reasonable ratio (0.5-1.5x)
   -> Success rate: 4.9%

   Examples:
     promo_price  new_unit_price test_remove_first_dot  ratio_to_unit
1948   1.119.976          119.00              1119.976       9.411563
1949   1.119.976          109.99              1119.976      10.182526
1950   1.119.976          109.99              1119.976      10.182526
1951   1.119.976          109.99              1119.976      10.182526
1952   1.119.976          109.99              1119.976      10.182526
1953   1.119.976          109.99              1119.976      10.182526
1954   1.119.976          109.99              1119.976      10.182526
1955   1.119.976          106.99              1119.976      10.468044
1956   1.119.976 

## Documentation

### Dataset Overview
- **19,326 rows** and **7 columns:** `sku`, `name`, `desc`, `price`, `promo_price`, `in_stock`, `type`
- 9,504 duplicates for sku / 9,777 duplicates for name / 14,373 duplicates for desc / 19,318 duplicates for type

### Missing Values
- sku: 0
- name: 0
- desc: 7
- price: 46
- promo_price: 0
- in_stock: 0
- type: 50

### Data Quality Notes
- The prices in `promo_price` appear to be incorrect - for 3-digit prices, the decimal point seems to have shifted one position to the right

### Summary Statistics (without data cleaning)
- **Average price:** $1,862.57 USD
- **Median:** $1,676.94 USD  
- **75th percentile:** $3,349 (three quarters of all products are below this value)